### PR TITLE
Fix wrong interpretation of AugAssign to AliasAssign

### LINKED ISSUE #1281: `AugAssign` interpreted as `AliasAssign`
### Describe the bug
When calling an `AugAssign` operator on a pointer, pyccel interprets this as an `AliasAssign`. This is probably due to the fact that `AugAssign` inherits from `Assign` and this difference is not taken into consideration when the `AliasAssign` is created. The problem exists in all languages so it must be a semantic problem.

### To Reproduce
Provide code to reproduce the behaviour:
```python
import numpy as np

if __name__ == '__main__':
    a = np.ones(5)
    b = a
    b += 3
```

### Error details
Here is the generated code in Fortran:
```fortran
program prog_prog_tmp4

  use tmp4

  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T , f64 => &
        C_DOUBLE
  implicit none

  real(f64), allocatable, target :: a(:)
  real(f64), pointer :: b(:)

  allocate(a(0:4_i64))
  a = 1.0_f64
  b(0:) => a
  b(0:) => 3_i64
  if (allocated(a)) then
    deallocate(a)
  end if

end program prog_prog_tmp4
```
C:
```c
#include "tmp4.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <stdint.h>
int main()
{
    t_ndarray a = {.shape = NULL};
    t_ndarray b = {.shape = NULL};
    a = array_create(1, (int64_t[]){INT64_C(5)}, nd_double, false);
    array_fill((double)1.0, a);
    alias_assign(&b, a);
    free_pointer(b);
    b = INT64_C(3);
    free_array(a);
    free_pointer(b);
    return 0;
}
```
and python:
```python
from numpy import ones

if __name__ == "__main__":
    a = ones((5), dtype=float)
    b = a
    b = 3
```

### Expected behaviour
The codes should be:
Fortran:
```fortran
program prog_prog_tmp4

  use tmp4

  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T , f64 => &
        C_DOUBLE
  implicit none

  real(f64), allocatable, target :: a(:)
  real(f64), pointer :: b(:)

  allocate(a(0:4_i64))
  a = 1.0_f64
  b(0:) => a
  b(0:) = b(0:) + 3_i64
  if (allocated(a)) then
    deallocate(a)
  end if

end program prog_prog_tmp4
```
C:
```c
#include "tmp4.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <stdint.h>
int main()
{
    t_ndarray a = {.shape = NULL};
    t_ndarray b = {.shape = NULL};
    a = array_create(1, (int64_t[]){INT64_C(5)}, nd_double, false);
    array_fill((double)1.0, a);
    alias_assign(&b, a);
    for (i = INT64_C(0); i < INT64_C(5); i += INT64_C(1))
    {
        GET_ELEMENT(b, nd_double, (int64_t)i) += INT64_C(3);
    }
    free_array(a);
    free_pointer(b);
    return 0;
}
```
and python:
```python
from numpy import ones

if __name__ == "__main__":
    a = ones((5), dtype=float)
    b = a
    b += 3
```

### PR DESCRIPTION
**Pull request by aihya**
Fixes #1281

This issue is solving the wrong interpretation of `AugAssign` into `AliasAssign`.

```python
import numpy as np

if __name__ == '__main__':
    a = np.ones(5)
    b = a
    b += 3
```
The code above, when converted to C, used to generate this:
```C
#include "test.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <stdint.h>
int main()
{
    t_ndarray a = {.shape = NULL};
    t_ndarray b = {.shape = NULL};
    a = array_create(1, (int64_t[]){INT64_C(5)}, nd_double, false);
    array_fill((double)1.0, a);
    alias_assign(&b, a);
    free_pointer(b);
    b = INT64_C(3);
    free_array(a);
    free_pointer(b);
    return 0;
}
```
Now it generates this:
```C
#include "test.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <stdint.h>
#include <stdio.h>
#include <inttypes.h>
int main()
{
    t_ndarray a = {.shape = NULL};
    t_ndarray b = {.shape = NULL};
    int64_t i;
    int64_t i_0001;
    a = array_create(1, (int64_t[]){INT64_C(5)}, nd_double, false);
    array_fill((double)1.0, a);
    alias_assign(&b, a);
    for (i = INT64_C(0); i < b.shape[INT64_C(0)]; i += INT64_C(1))
    {
        GET_ELEMENT(b, nd_double, (int64_t)i) += INT64_C(3);
    }
    free_array(a);
    free_pointer(b);
    return 0;
}
```
which removes an unnecessary de-allocation of `b` after the call to `alias_assign(&b, a)` and replaces `b = INT64_C(3)` with for loop to broadcast `+=` operation on all elements of `b`, which point on `a`.

For Fortran:
```fortran
program prog_prog_test

  use test_0001

  use, intrinsic :: ISO_C_Binding, only : f64 => C_DOUBLE , i64 => &
        C_INT64_T
  implicit none

  real(f64), allocatable, target :: a(:)
  real(f64), pointer :: b(:)

  allocate(a(0:4_i64))
  a = 1.0_f64
  b(0:) => a
  b(0:) => 3_i64
  if (allocated(a)) then
    deallocate(a)
  end if

end program prog_prog_test
```
the line `b(0:) => 3_i64` is replaced with `b = b + 3_i64`

### CODE REVIEW

**Comment by aihya** [comment]
@EmilyBourne the fix I did seems to be working fine for linux. For some reason there's an `ImportError` raised for both windows and macos and it doesn't look it has any relationship to the fix in anyway.

**Comment by EmilyBourne** [comment]
> @EmilyBourne the fix I did seems to be working fine for linux. For some reason there's an `ImportError` raised for both windows and macos and it doesn't look it has any relationship to the fix in anyway.

The MacOS and Windows errors are on all branches. See issue #1272 and the associated PR. We're just waiting for someone with merge rights to fix it ;)

File: tests/epyccel/test_epyccel_augassign.py
Comment (EmilyBourne) [inline tests/epyccel/test_epyccel_augassign.py:28]: It is not necessary to specify the  value that you think will be returned the pure python execution of the function is there to obtain this value.

By the way you need:
```python
    x1_int     = np.zeros(5, dtype=int)
    x1_float   = np.zeros(5, dtype=float)
    x1_complex = np.zeros(5, dtype=complex)
    x2_int     = np.zeros(5, dtype=int)
    x2_float   = np.zeros(5, dtype=float)
    x2_complex = np.zeros(5, dtype=complex)
```

to obtain the result you expect.

It would also be good to check that the pointer is working correctly and that the argument has indeed changed

Comment (aihya) [reply tests/epyccel/test_epyccel_augassign.py]: > It would also be good to check that the pointer is working correctly and that the argument has indeed changed
What pointer? I don't see what you mean.

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_augassign.py]:
```
b = a
```
is not a copy. `b` points at `a`. So when `b` is modified, `a` should also be modified. It would be good to check that:
```python
np.all_equal(x1_int, x2_int)
```

**Comment by EmilyBourne** [comment]
@aihya I see you have marked the PR as ready for review, but have not added any labels. Please see the [review process docs](https://github.com/pyccel/pyccel/blob/master/developer_docs/review_process.md#review-stage-labels) for details about how PRs are managed in pyccel

File: pyccel/parser/semantic.py
Comment (Pinkyboi) [inline pyccel/parser/semantic.py:1216]: Maybe i'm missing something but why do we need to free aliases ?

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1216]: Aliases contain shape and strides arrays which are independent of the original array. This allows you to do:
```
a = np.ones(4,6)
b = a[:2,:3]
a.shape # 4,6
b.shape # 2,2
```
These arrays need freeing when the alias goes out of scope

Comment (Pinkyboi) [reply pyccel/parser/semantic.py:1216]: Oh i see so we free the shape and strides in that case makes sense.

Comment (Pinkyboi) [reply pyccel/parser/semantic.py:1216]: Well it's seems fine to me then.

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1216]: In that case, you should `approve` the PR

**Review by Pinkyboi — APPROVED** [review]

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:2641]: Why do you duplicate this line of code:
```python
new_expr = AugAssign(l, expr.op, r)
```
rather than re-ordering the `if`?
```python
if isinstance(expr, AugAssign):
    new_expr = AugAssign(l, expr.op, r)
elif is_pointer_i:
    new_expr = AliasAssign(l, r)
```

File: tests/epyccel/test_epyccel_augassign.py
Comment (EmilyBourne) [inline tests/epyccel/test_epyccel_augassign.py:35]: Why do you not test the output of the function as well as the argument?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: tests/epyccel/test_epyccel_augassign.py
Comment (aihya) [reply tests/epyccel/test_epyccel_augassign.py:35]: I don't think there is a need for that, `np.array_equal` checks the elements of the arrays.

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_augassign.py:35]: But `a` and `b` are not the same object. When everything is working correctly I agree that `np.array_equal` is enough, but if something is broken, having both allows us to tell the difference between a broken `AugAssign` and a broken pointer

File: tests/epyccel/modules/augassign.py
Comment (ohachim) [inline tests/epyccel/modules/augassign.py:99]: Checking the first elements is enough?

Comment (ohachim) [reply tests/epyccel/modules/augassign.py:99]: nvm, you are using `array_equal`, but that means the return is not necessary?

File: pyccel/parser/semantic.py
Comment (aihya) [reply pyccel/parser/semantic.py]: fixed https://github.com/pyccel/pyccel/pull/1286/commits/0345a8d7484f597364948d4d65eff6a70405ff2a https://github.com/pyccel/pyccel/pull/1286/commits/f8459aa46b3f9da260c48f8c45604567712663ad

File: tests/epyccel/modules/augassign.py
Comment (aihya) [reply tests/epyccel/modules/augassign.py:99]: It's use now in the test functions: https://github.com/pyccel/pyccel/pull/1286/commits/7f2400caf3017a9be96cca9df9f72595526e50ac

File: tests/epyccel/test_epyccel_augassign.py
Comment (aihya) [reply tests/epyccel/test_epyccel_augassign.py]: https://github.com/pyccel/pyccel/pull/1286/commits/7f2400caf3017a9be96cca9df9f72595526e50ac

**Review by EmilyBourne — APPROVED** [review]

**Review by ohachim — APPROVED** [review]

**Review by yguclu — APPROVED** [review]
Well done!

**Pull request merged**
